dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AbstractDatastoreController.java

```diff
@@ -0,0 +1,127 @@
+import static org.hisp.dhis.webapi.utils.ContextUtils.setNoStore;
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import javax.servlet.http.HttpServletResponse;
+import lombok.Value;
+import org.hisp.dhis.common.OpenApi;
+import org.hisp.dhis.datastore.DatastoreFields;
+import org.hisp.dhis.datastore.DatastoreQuery;
+import org.hisp.dhis.datastore.DatastoreQuery.Field;
+import org.hisp.dhis.dxf2.webmessage.WebMessageUtils;
+import org.hisp.dhis.feedback.ConflictException;
+import org.hisp.dhis.webapi.JsonWriter;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public abstract class AbstractDatastoreController {
+
+  @Autowired private ObjectMapper jsonMapper;
+
+  @Value
+  @OpenApi.Shared(value = false)
+  static class Pager {
+    int page;
+    int pageSize;
+  }
+
+  @Value
+  static class EntriesResponse {
+
+    Pager pager;
+    List<Map<String, JsonNode>> entries;
+  }
+
+  @FunctionalInterface
+  interface DatastoreQueryExecutor {
+    boolean getEntries(DatastoreQuery query, Predicate<Stream<DatastoreFields>> transform)
+        throws ConflictException;
+  }
+
+  void writeEntries(
+      HttpServletResponse response, DatastoreQuery query, DatastoreQueryExecutor runQuery)
+      throws IOException {
+    response.setContentType(APPLICATION_JSON_VALUE);
+    setNoStore(response);
+
+    try (PrintWriter writer = response.getWriter();
+        JsonWriter out = new JsonWriter(writer)) {
+      try {
+        List<String> members = query.getFields().stream().map(Field::getAlias).collect(toList());
+        runQuery.getEntries(
+            query,
+            entries -> {
+              if (!query.isHeadless()) {
+                writer.write("{\"pager\":{");
+                writer.write("\"page\":" + query.getPage() + ",");
+                writer.write("\"pageSize\":" + query.getPageSize());
+                writer.write("},\"entries\":");
+              }
+              out.writeEntries(members, entries);
+              return true;
+            });
+        if (!query.isHeadless()) {
+          writer.write("}");
+        }
+      } catch (RuntimeException | ConflictException ex) {
+        response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+        Throwable cause = ex.getCause();
+        String msg =
+            "Unknown error when running the query: "
+                + (cause != null && ex.getMessage().contains("could not extract ResultSet")
+                    ? cause.getMessage()
+                    : ex.getMessage());
+        if (cause != null
+            && cause.getMessage().contains("cannot cast type ")
+            && cause.getMessage().contains(" to double precision")) {
```

Comment:
going above and beyond to output a good error message which helps, nice.

Reply:
Haha, I moved that from the `DatastoreController` without noticing it. I did good when I was implementing it originally :)